Netkan/Services/CachingHttpService.cs

```diff
@@ -0,0 +1,141 @@
+
+            _curl = new CurlEasy { UserAgent = CKAN.Net.UserAgentString };
+
+            var caBundle = ResolveCurlCaBundle();
+            if (caBundle != null)
+            {
+                _curl.CaInfo = caBundle;
+            }
+        }
+
+        public string DownloadPackage(Uri url, string identifier)
+        {
+            EnsureNotDisposed();
+
+            return _cache.GetCachedFilename(url) ?? 
+                _cache.Store(url, CKAN.Net.Download(url), string.Format("netkan-{0}", identifier), move: true);
+        }
+
+        public string DownloadText(Uri url)
+        {
+            EnsureNotDisposed();
+
+            Log.DebugFormat("About to download {0}", url);
+
+            var content = string.Empty;
+
+            _curl.Url = url.ToString();
+            _curl.WriteData = null;
+            _curl.WriteFunction = delegate(byte[] buf, int size, int nmemb, object extraData)
+            {
+                content += Encoding.UTF8.GetString(buf);
+                return size * nmemb;
+            };
+
+            var result = _curl.Perform();
+
+            if (result != CurlCode.Ok)
+            {
+                throw new Exception("Curl download failed with error " + result);
+            }
+
+            Log.DebugFormat("Download from {0}:\n\n{1}", url, content);
+
+            return content;
+        }
+
+        /// <summary>
+        /// Resolves the location of the cURL CA bundle file to use.
+        /// </summary>
+        /// <returns>The absolute file path to the bundle file or null if none is found.</returns>
+        private static string ResolveCurlCaBundle()
+        {
+            const string caBundleFileName = "curl-ca-bundle.crt";
+            const string ckanSubDirectoryName = "CKAN";
+
+            var bundle = new[]
+            {
+                // Working Directory
+                Environment.CurrentDirectory,
+
+                // Executable Directory
+                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
+
+                // %LOCALAPPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), ckanSubDirectoryName),
+
+                // %APPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), ckanSubDirectoryName),
+
+                // %PROGRAMDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), ckanSubDirectoryName),
+            }
+            .Select(i => Path.Combine(i, caBundleFileName))
+            .FirstOrDefault(File.Exists);
+
+            Log.InfoFormat("Using curl-ca bundle: {0}",bundle ?? "(none)");
+
+            return bundle;
+        }
+    }
+
+    internal sealed partial class CachingHttpService
+    {
+        private bool _isDisposed;
+
+        ~CachingHttpService()
+        {
+            Dispose(disposing: false);
```

Comment:
So two scenarios here:

##### Proper Disposal
In this scenario the object is constructed and `Dispose()` is called on it, either manually or automatically by the termination of a `using` block. The order of events is then:

- `Dispose()` is called
- `Dispose(true)` is then called
- We check to see if we were called by the `Dispose()` method by checking the `disposing` flag. This is `true`, so dispose the managed objects we 'own'.
- Regardless of the value of `disposing`, cleanup unmanaged resources. *We don't have any so do nothing*.
- At some indeterminate time in the future, the object will be garbage collected. Just before it's destroyed, the finalizer will check to see if the object has a destructor, it does. It will then check to see if finalization should be suppressed, we called `GC.SuppressFinalize(this)`, so yes. The finalizer destroys the object will no further code execution.

##### Improper Disposal
In this scenario is object is constructed and `Dispose()` is never called on it. At some point in the future all references to it are lost and the object becomes garbage. During garbage collection:

- The finalizer checks if the object has a destructor, it does.
- The finalizer checks if finalization should be suppressed. In this scenario `GC.SuppressFinalize(this)` is never called, so no.
- The finalizer executes the destructor.
- `Dispose(false)` is executed.
- We check to see if we were called by the `Dispose()` method by checking the `disposing` flag. This is `false`, so ignore any managed objects. Why? Because during finalization typical reference rules no longer apply. Any managed objects we hold a reference to could already be finalized and references we hold to them would be `null`, otherwise it would be impossible to garbage collect circular references. We could perform a null check before accessing them, but those objects could have their own references nulled out and they would be in an indeterminate state they did not expect to be in, leading to all sorts of potential exceptions being thrown (and throwing exceptions on the finalizer thread == very bad). Therefore standard procedure is to ignore managed objects and assume their own destructors will handle clean-up.
- Regardless of the value of `disposing`, cleanup unmanaged resources. *We don't have any so do nothing*.

So basically, this scenario is a degenerate case. Nothing happens except `_isDisposed` gets set to `true`. If we happened to have unmanaged resources, we would clean them up, but we don't.

Why bother with a `Dispose()` method at all? Why not just let the destructors handle all clean-up?
- Predictability, you know exactly when the clean-up code will execute, instead of an effectively random time in the future.
- Performance, when everything is disposed properly, destructors should never be executed and the only code running on the finalizer thread is the actual garbage collection.
- Robustness, since `Dispose()` gets called in normal code you can make safe assumptions about the state of the world, versus on the finalizer thread where some objects could still be live, some objects could be destroyed and you can't predict which are which.